File: src/utils/dynamodb.py
"""
Utilidades para interactuar con DynamoDB
"""
import os
import boto3
from typing import Any, Dict, List, Optional
from decimal import Decimal
from boto3.dynamodb.conditions import Key, Attr


# Inicializar cliente DynamoDB
dynamodb = boto3.resource('dynamodb')

# Nombres de tablas desde variables de entorno
USERS_TABLE = os.environ.get('USERS_TABLE')
INCIDENTS_TABLE = os.environ.get('INCIDENTS_TABLE')
CONNECTIONS_TABLE = os.environ.get('CONNECTIONS_TABLE')


def get_table(table_name: str):
    """
    Obtiene una referencia a una tabla de DynamoDB
    
    Args:
        table_name: Nombre de la tabla
    
    Returns:
        Objeto Table de boto3
    """
    return dynamodb.Table(table_name)


def put_item(table_name: str, item: Dict[str, Any]) -> Dict[str, Any]:
    """
    Inserta un item en DynamoDB
    
    Args:
        table_name: Nombre de la tabla
        item: Item a insertar
    
    Returns:
        Respuesta de DynamoDB
    """
    table = get_table(table_name)
    return table.put_item(Item=item)


def get_item(table_name: str, key: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """
    Obtiene un item de DynamoDB por su clave primaria
    
    Args:
        table_name: Nombre de la tabla
        key: Clave primaria del item
    
    Returns:
        Item encontrado o None
    """
    table = get_table(table_name)
    response = table.get_item(Key=key)
    return response.get('Item')


def update_item(
    table_name: str,
    key: Dict[str, Any],
    update_expression: str,
    expression_values: Dict[str, Any],
    expression_names: Optional[Dict[str, str]] = None,
    return_values: str = "ALL_NEW"
) -> Dict[str, Any]:
    """
    Actualiza un item en DynamoDB
    
    Args:
        table_name: Nombre de la tabla
        key: Clave primaria del item
        update_expression: Expresión de actualización
        expression_values: Valores para la expresión
        expression_names: Nombres de atributos (opcional)
        return_values: Qué valores retornar
    
    Returns:
        Item actualizado
    """
    table = get_table(table_name)
    
    params = {
        'Key': key,
        'UpdateExpression': update_expression,
        'ExpressionAttributeValues': expression_values,
        'ReturnValues': return_values
    }
    
    if expression_names:
        params['ExpressionAttributeNames'] = expression_names
    
    response = table.update_item(**params)
    return response.get('Attributes')


def delete_item(table_name: str, key: Dict[str, Any]) -> Dict[str, Any]:
    """
    Elimina un item de DynamoDB
    
    Args:
        table_name: Nombre de la tabla
        key: Clave primaria del item
    
    Returns:
        Respuesta de DynamoDB
    """
    table = get_table(table_name)
    return table.delete_item(Key=key)


def query_items(
    table_name: str,
    key_condition: Any,
    index_name: Optional[str] = None,
    filter_expression: Optional[Any] = None,
    limit: Optional[int] = None,
    scan_forward: bool = True,
    last_evaluated_key: Optional[Dict] = None
) -> Dict[str, Any]:
    """
    Consulta items en DynamoDB con condiciones
    
    Args:
        table_name: Nombre de la tabla
        key_condition: Condición de clave (usando Key de boto3)
        index_name: Nombre del GSI (opcional)
        filter_expression: Expresión de filtro adicional
        limit: Límite de items
        scan_forward: True para orden ascendente, False para descendente
        last_evaluated_key: Para paginación
    
    Returns:
        Dict con Items y LastEvaluatedKey
    """
    table = get_table(table_name)
    
    params = {
        'KeyConditionExpression': key_condition,
        'ScanIndexForward': scan_forward
    }
    
    if index_name:
        params['IndexName'] = index_name
    
    if filter_expression is not None:
        params['FilterExpression'] = filter_expression
    
    if limit:
        params['Limit'] = limit
    
    if last_evaluated_key:
        params['ExclusiveStartKey'] = last_evaluated_key
    
    response = table.query(**params)
    
    return {
        'items': response.get('Items', []),
        'lastEvaluatedKey': response.get('LastEvaluatedKey'),
        'count': response.get('Count', 0)
    }


def scan_items(
    table_name: str,
    filter_expression: Optional[Any] = None,
    limit: Optional[int] = None,
    last_evaluated_key: Optional[Dict] = None
) -> Dict[str, Any]:
    """
    Escanea toda una tabla (usar con precaución)
    
    Args:
        table_name: Nombre de la tabla
        filter_expression: Expresión de filtro
        limit: Límite de items
        last_evaluated_key: Para paginación
    
    Returns:
        Dict con Items y LastEvaluatedKey
    """
    table = get_table(table_name)
    
    params = {}
    
    if filter_expression is not None:
        params['FilterExpression'] = filter_expression
    
    if limit:
        params['Limit'] = limit
    
    if last_evaluated_key:
        params['ExclusiveStartKey'] = last_evaluated_key
    
    response = table.scan(**params)
    
    return {
        'items': response.get('Items', []),
        'lastEvaluatedKey': response.get('LastEvaluatedKey'),
        'count': response.get('Count', 0)
    }


def batch_get_items(table_name: str, keys: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Obtiene múltiples items en una sola operación
    
    Args:
        table_name: Nombre de la tabla
        keys: Lista de claves primarias
    
    Returns:
        Lista de items encontrados
    """
    if not keys:
        return []
    
    response = dynamodb.batch_get_item(
        RequestItems={
            table_name: {
                'Keys': keys
            }
        }
    )
    
    return response.get('Responses', {}).get(table_name, [])


def batch_write_items(table_name: str, items: List[Dict[str, Any]]) -> bool:
    """
    Escribe múltiples items en una sola operación
    
    Args:
        table_name: Nombre de la tabla
        items: Lista de items a insertar
    
    Returns:
        True si fue exitoso
    """
    if not items:
        return True
    
    table = get_table(table_name)
    
    with table.batch_writer() as batch:
        for item in items:
            batch.put_item(Item=item)
    
    return True


def decimal_to_float(obj: Any) -> Any:
    """
    Convierte objetos Decimal a float recursivamente
    Útil para serializar respuestas de DynamoDB a JSON
    
    Args:
        obj: Objeto que puede contener Decimals
    
    Returns:
        Objeto con Decimals convertidos a float
    """
    if isinstance(obj, list):
        return [decimal_to_float(i) for i in obj]
    elif isinstance(obj, dict):
        return {k: decimal_to_float(v) for k, v in obj.items()}
    elif isinstance(obj, Decimal):
        if obj % 1 == 0:
            return int(obj)
        return float(obj)
    return obj


# Funciones específicas para el sistema de incidentes

def get_user_by_id(user_id: str) -> Optional[Dict[str, Any]]:
    """Obtiene un usuario por su ID"""
    return get_item(USERS_TABLE, {'userId': user_id})


def get_user_by_email(email: str) -> Optional[Dict[str, Any]]:
    """Obtiene un usuario por su email"""
    result = query_items(
        USERS_TABLE,
        Key('email').eq(email),
        index_name='EmailIndex',
        limit=1
    )
    items = result.get('items', [])
    return items[0] if items else None


def get_incident_by_id(incident_id: str) -> Optional[Dict[str, Any]]:
    """Obtiene un incidente por su ID"""
    return get_item(INCIDENTS_TABLE, {'incidentId': incident_id})


def get_incidents_by_status(status: str, limit: int = 50, last_key: Optional[Dict] = None) -> Dict[str, Any]:
    """Obtiene incidentes por estado"""
    return query_items(
        INCIDENTS_TABLE,
        Key('status').eq(status),
        index_name='StatusCreatedAtIndex',
        limit=limit,
        scan_forward=False,
        last_evaluated_key=last_key
    )


def get_incidents_by_worker(worker_id: str, limit: int = 50) -> Dict[str, Any]:
    """Obtiene incidentes asignados a un trabajador"""
    return query_items(
        INCIDENTS_TABLE,
        Key('assignedTo').eq(worker_id),
        index_name='AssignedToIndex',
        limit=limit,
        scan_forward=False
    )


def get_workers() -> List[Dict[str, Any]]:
    """Obtiene todos los trabajadores (workers)"""
    result = query_items(
        USERS_TABLE,
        Key('role').eq('worker'),
        index_name='RoleIndex'
    )
    return result.get('items', [])

File: src/utils/dynamodb.js
/**
 * Utilidades para interactuar con DynamoDB
 */
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { 
  DynamoDBDocumentClient, 
  PutCommand, 
  GetCommand, 
  UpdateCommand, 
  DeleteCommand,
  QueryCommand,
  ScanCommand,
  BatchGetCommand,
  BatchWriteCommand
} from '@aws-sdk/lib-dynamodb';

// Inicializar cliente DynamoDB
const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);

// Nombres de tablas desde variables de entorno
const USERS_TABLE = process.env.USERS_TABLE;
const INCIDENTS_TABLE = process.env.INCIDENTS_TABLE;
const CONNECTIONS_TABLE = process.env.CONNECTIONS_TABLE;

/**
 * Inserta un item en DynamoDB
 * 
 * @param {string} tableName - Nombre de la tabla
 * @param {Object} item - Item a insertar
 * @returns {Promise<Object>} Respuesta de DynamoDB
 */
export async function putItem(tableName, item) {
  const command = new PutCommand({
    TableName: tableName,
    Item: item
  });
  return docClient.send(command);
}

/**
 * Obtiene un item de DynamoDB por su clave primaria
 * 
 * @param {string} tableName - Nombre de la tabla
 * @param {Object} key - Clave primaria del item
 * @returns {Promise<Object|null>} Item encontrado o null
 */
export async function getItem(tableName, key) {
  const command = new GetCommand({
    TableName: tableName,
    Key: key
  });
  const response = await docClient.send(command);
  return response.Item || null;
}

/**
 * Actualiza un item en DynamoDB
 * 
 * @param {string} tableName - Nombre de la tabla
 * @param {Object} key - Clave primaria del item
 * @param {string} updateExpression - Expresión de actualización
 * @param {Object} expressionValues - Valores para la expresión
 * @param {Object} [expressionNames=null] - Nombres de atributos (opcional)
 * @param {string} [returnValues='ALL_NEW'] - Qué valores retornar
 * @returns {Promise<Object>} Item actualizado
 */
export async function updateItem(
  tableName,
  key,
  updateExpression,
  expressionValues,
  expressionNames = null,
  returnValues = 'ALL_NEW'
) {
  const params = {
    TableName: tableName,
    Key: key,
    UpdateExpression: updateExpression,
    ExpressionAttributeValues: expressionValues,
    ReturnValues: returnValues
  };

  if (expressionNames) {
    params.ExpressionAttributeNames = expressionNames;
  }

  const command = new UpdateCommand(params);
  const response = await docClient.send(command);
  return response.Attributes;
}

/**
 * Elimina un item de DynamoDB
 * 
 * @param {string} tableName - Nombre de la tabla
 * @param {Object} key - Clave primaria del item
 * @returns {Promise<Object>} Respuesta de DynamoDB
 */
export async function deleteItem(tableName, key) {
  const command = new DeleteCommand({
    TableName: tableName,
    Key: key
  });
  return docClient.send(command);
}

/**
 * Consulta items en DynamoDB con condiciones
 * 
 * @param {string} tableName - Nombre de la tabla
 * @param {string} keyConditionExpression - Expresión de condición de clave
 * @param {Object} expressionValues - Valores para la expresión
 * @param {Object} options - Opciones adicionales
 * @returns {Promise<Object>} Objeto con items, lastEvaluatedKey y count
 */
export async function queryItems(tableName, keyConditionExpression, expressionValues, options = {}) {
  const params = {
    TableName: tableName,
    KeyConditionExpression: keyConditionExpression,
    ExpressionAttributeValues: expressionValues,
    ScanIndexForward: options.scanForward !== false
  };

  if (options.indexName) {
    params.IndexName = options.indexName;
  }

  if (options.filterExpression) {
    params.FilterExpression = options.filterExpression;
  }

  if (options.limit) {
    params.Limit = options.limit;
  }

  if (options.lastEvaluatedKey) {
    params.ExclusiveStartKey = options.lastEvaluatedKey;
  }

  if (options.expressionNames) {
    params.ExpressionAttributeNames = options.expressionNames;
  }

  const command = new QueryCommand(params);
  const response = await docClient.send(command);

  return {
    items: response.Items || [],
    lastEvaluatedKey: response.LastEvaluatedKey,
    count: response.Count || 0
  };
}

/**
 * Escanea toda una tabla (usar con precaución)
 * 
 * @param {string} tableName - Nombre de la tabla
 * @param {Object} options - Opciones de scan
 * @returns {Promise<Object>} Objeto con items, lastEvaluatedKey y count
 */
export async function scanItems(tableName, options = {}) {
  const params = {
    TableName: tableName
  };

  if (options.filterExpression) {
    params.FilterExpression = options.filterExpression;
  }

  if (options.expressionValues) {
    params.ExpressionAttributeValues = options.expressionValues;
  }

  if (options.limit) {
    params.Limit = options.limit;
  }

  if (options.lastEvaluatedKey) {
    params.ExclusiveStartKey = options.lastEvaluatedKey;
  }

  const command = new ScanCommand(params);
  const response = await docClient.send(command);

  return {
    items: response.Items || [],
    lastEvaluatedKey: response.LastEvaluatedKey,
    count: response.Count || 0
  };
}

/**
 * Obtiene múltiples items en una sola operación
 * 
 * @param {string} tableName - Nombre de la tabla
 * @param {Array<Object>} keys - Lista de claves primarias
 * @returns {Promise<Array>} Lista de items encontrados
 */
export async function batchGetItems(tableName, keys) {
  if (!keys || keys.length === 0) {
    return [];
  }

  const command = new BatchGetCommand({
    RequestItems: {
      [tableName]: {
        Keys: keys
      }
    }
  });

  const response = await docClient.send(command);
  return response.Responses?.[tableName] || [];
}

/**
 * Escribe múltiples items en una sola operación
 * 
 * @param {string} tableName - Nombre de la tabla
 * @param {Array<Object>} items - Lista de items a insertar
 * @returns {Promise<boolean>} True si fue exitoso
 */
export async function batchWriteItems(tableName, items) {
  if (!items || items.length === 0) {
    return true;
  }

  // DynamoDB permite máximo 25 items por batch
  const batches = [];
  for (let i = 0; i < items.length; i += 25) {
    batches.push(items.slice(i, i + 25));
  }

  for (const batch of batches) {
    const command = new BatchWriteCommand({
      RequestItems: {
        [tableName]: batch.map(item => ({
          PutRequest: { Item: item }
        }))
      }
    });
    await docClient.send(command);
  }

  return true;
}

// Funciones específicas para el sistema de incidentes

/**
 * Obtiene un usuario por su ID
 * @param {string} userId - ID del usuario
 * @returns {Promise<Object|null>}
 */
export async function getUserById(userId) {
  return getItem(USERS_TABLE, { userId });
}

/**
 * Obtiene un usuario por su email
 * @param {string} email - Email del usuario
 * @returns {Promise<Object|null>}
 */
export async function getUserByEmail(email) {
  const result = await queryItems(
    USERS_TABLE,
    'email = :email',
    { ':email': email },
    { indexName: 'EmailIndex', limit: 1 }
  );
  const items = result.items || [];
  return items.length > 0 ? items[0] : null;
}

/**
 * Obtiene un incidente por su ID
 * @param {string} incidentId - ID del incidente
 * @returns {Promise<Object|null>}
 */
export async function getIncidentById(incidentId) {
  return getItem(INCIDENTS_TABLE, { incidentId });
}

/**
 * Obtiene incidentes por estado
 * @param {string} status - Estado del incidente
 * @param {number} limit - Límite de resultados
 * @param {Object} lastKey - Última clave evaluada para paginación
 * @returns {Promise<Object>}
 */
export async function getIncidentsByStatus(status, limit = 50, lastKey = null) {
  return queryItems(
    INCIDENTS_TABLE,
    '#status = :status',
    { ':status': status },
    { 
      indexName: 'StatusCreatedAtIndex',
      limit,
      scanForward: false,
      lastEvaluatedKey: lastKey,
      expressionNames: { '#status': 'status' }
    }
  );
}

/**
 * Obtiene incidentes asignados a un trabajador
 * @param {string} workerId - ID del trabajador
 * @param {number} limit - Límite de resultados
 * @returns {Promise<Object>}
 */
export async function getIncidentsByWorker(workerId, limit = 50) {
  return queryItems(
    INCIDENTS_TABLE,
    'assignedTo = :workerId',
    { ':workerId': workerId },
    { 
      indexName: 'AssignedToIndex',
      limit,
      scanForward: false
    }
  );
}

/**
 * Obtiene todos los trabajadores (workers)
 * @returns {Promise<Array>}
 */
export async function getWorkers() {
  const result = await queryItems(
    USERS_TABLE,
    '#role = :role',
    { ':role': 'worker' },
    { 
      indexName: 'RoleIndex',
      expressionNames: { '#role': 'role' }
    }
  );
  return result.items || [];
}